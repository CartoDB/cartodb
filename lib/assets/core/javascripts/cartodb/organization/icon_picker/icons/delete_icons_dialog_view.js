var BaseDialog = require('../../../../../javascripts/cartodb/common/views/base_dialog/view');

module.exports = BaseDialog.extend({

  initialize: function () {
    if (!this.options.okCallback) {
      throw new Error('Callback for OK action is mandatory.');
    }
    this.elder('initialize');
    this._numOfIcons = this.options.numOfIcons || 0;
    this._okCallback = this.options.okCallback;
  },

  render_content: function () {
    return this.getTemplate('organization/icon_picker/icons/delete_icons_modal')({
      numOfIcons: this._numOfIcons
    });
  },

  ok: function () {
    this._okCallback();
    this.close();
  }
});
