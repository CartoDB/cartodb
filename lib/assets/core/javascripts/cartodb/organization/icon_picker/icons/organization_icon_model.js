var Backbone = require('backbone-cdb-v3');
var _ = require('underscore-cdb-v3');

module.exports = Backbone.Model.extend({

  defaults: {
    selected: false,
    visible: false,
    deleted: false
  },

  fileAttribute: 'resource',

  save: function (attrs, options) {
    options || (options = {});
    attrs || (attrs = _.clone(this.attributes));

    // Filter the data to send to the server
    attrs = _.omit(attrs, ['selected', 'visible', 'deleted']);
    options.data = JSON.stringify(attrs);

    // Proxy the call to the original save function
    return Backbone.Model.prototype.save.call(this, attrs, options);
  }
});
