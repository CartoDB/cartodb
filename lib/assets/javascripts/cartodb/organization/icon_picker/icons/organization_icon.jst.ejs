<div class="IconItem-icon js-asset">
  <img height="24" src="<%- url %>" alt="icon" crossorigin="anonymous" />
</div>
