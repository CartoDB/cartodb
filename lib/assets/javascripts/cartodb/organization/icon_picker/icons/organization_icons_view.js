var _ = require('underscore-cdb-v3');
var cdb = require('cartodb.js-v3');
var IconCollection = require('./organization_icon_collection');
var IconView = require('./organization_icon_view');
var DeleteIconsDialog = require('./delete_icons_dialog_view');

module.exports = cdb.core.View.extend({

  events: {
    'click .js-addIcon': '_onAddIconClicked',
    'click .js-selectAllIcons': '_onSelectAllIconsClicked',
    'click .js-deselectAllIcons': '_onDeselectAllIconsClicked',
    'click .js-deleteIcons': '_onDeleteIconsClicked',
    'change .js-inputFile': '_onFileSelected'
  },

  initialize: function () {
    if (!this.options.orgId) { throw new Error('Organization ID is required.'); }
    this._orgId = this.options.orgId;

    this.template = cdb.templates.getTemplate('organization/icon_picker/icons/organization_icons_template');
    this.model = new cdb.core.Model({
      isProcessRunning: false
    });
    this._iconCollection = new IconCollection(null, {
      orgId: this._orgId
    });
    this._numOfUploadingProcesses = 0;
    this._numOfDeletingProcesses = 0;
    this._fetchErrorMessage = 'Error fetching organization icons. Please refresh the page.';
    this._runningMessage = '';
    this.render();
    this._fetchAllIcons();
    this._initBinds();
  },

  render: function () {
    this.$el.html(this.template());

    return this;
  },

  _initBinds: function () {
    this._iconCollection.on('change:selected', this._refreshActions, this);
    this.model.on('change:isProcessRunning', this._onProcessRunningChanged, this);
    this.add_related_model(this._iconCollection);
  },

  _fetchAllIcons: function () {
    this._iconCollection.fetch({
      success: this._renderAllIcons.bind(this),
      error: this._onFetchIconsError.bind(this)
    });
  },

  _renderAllIcons: function () {
    _.each(this._iconCollection.models, function (icon) {
      this._addIconElement(icon);
    }, this);
  },

  _onFetchIconsError: function () {
    this._showErrorMessage(this._fetchErrorMessage);
  },

  _renderIcon: function (iconModel) {
    var iconView = new IconView({
      model: iconModel
    });
    iconView.render();
    iconModel.set('visible', true);
    this.$('.js-items').append(iconView.$el);
  },

  _addIconElement: function (iconModel) {
    this._renderIcon(iconModel);
  },

  _onAddIconClicked: function (event) {
    this.killEvent(event);

    this._hideErrorMessage();
    this.$('.js-inputFile').trigger('click');
  },

  _parseResponseText: function (response) {
    if (response && response.responseText) {
      try {
        var text = JSON.parse(response.responseText);
        if (text && text.errors && typeof text.errors === 'string') {
          return text.errors;
        }
      } catch (exc) {
        // Swallow
      }
    }
    return '';
  },

  _getSelectedFiles: function () {
    return this.$('.js-inputFile').prop('files');
  },

  _onFileSelected: function () {
    var files = this._getSelectedFiles();

    _.each(files, function (file) {
      this._iconCollection.create({
        kind: 'organization_asset',
        resource: file
      }, {
        beforeSend: this._beforeIconUpload.bind(this),
        success: this._onIconUploaded.bind(this),
        error: this._onIconUploadError.bind(this),
        complete: this._onIconUploadComplete.bind(this)
      });
    }, this);
  },

  _beforeIconUpload: function () {
    this._numOfUploadingProcesses++;
    if (this._numOfUploadingProcesses > 0) {
      this._runningMessage = 'Uploading icons...';
      this.model.set('isProcessRunning', true);
    }
  },

  _onIconUploaded: function (iconModel) {
    this._resetFileSelection();
    this._addIconElement(iconModel);
  },

  _onIconUploadError: function (model, response) {
    var errorText = this._parseResponseText(response);
    this._resetFileSelection();
    this._showErrorMessage(this._uploadErrorMessage(errorText));
  },

  _onIconUploadComplete: function () {
    this._numOfUploadingProcesses--;
    if (this._numOfUploadingProcesses <= 0) {
      this.model.set('isProcessRunning', false);
    }
  },

  _resetFileSelection: function () {
    this.$('.js-inputFile').val('');
  },

  _show: function (selector) {
    this.$(selector).removeClass('is-hidden');
  },

  _hide: function (selector) {
    this.$(selector).addClass('is-hidden');
  },

  _refreshActions: function () {
    if (this.model.get('isProcessRunning')) {
      return;
    }
    var limit = Math.min(this._iconCollection.length);
    var numOfSelectedIcons = this._getNumberOfSelectedIcons();
    var iconText = (numOfSelectedIcons === 1 ? '1 icon selected' : '' + numOfSelectedIcons + ' icons selected');
    this.$('.js-iconMainLabel').text(iconText);

    if (numOfSelectedIcons === 0) {
      this.$('.js-iconMainLabel').text('');
      this._hide('.js-iconMainLabel, .js-selectAllIcons, .js-deselectAllIcons, .js-deleteIcons');
      this._show('.js-iconsInfo');
    } else if (numOfSelectedIcons < limit) {
      this._show('.js-iconMainLabel, .js-selectAllIcons, .js-deleteIcons');
      this._hide('.js-deselectAllIcons, .js-iconsInfo');
    } else {
      this._show('.js-iconMainLabel, .js-deselectAllIcons, .js-deleteIcons');
      this._hide('.js-selectAllIcons, .js-iconsInfo');
    }

    if (numOfSelectedIcons > 1) {
      this.$('.js-deleteIcons a').text('Delete icons...');
    } else if (numOfSelectedIcons === 1) {
      this.$('.js-deleteIcons a').text('Delete icon...');
    }
  },

  _hideActions: function () {
    this._hide('.js-selectAllIcons, .js-deselectAllIcons, .js-deleteIcons, .js-iconsInfo');
  },

  _onDeselectAllIconsClicked: function (event) {
    event.preventDefault();
    this._iconCollection.each(function (icon) {
      icon.set('selected', false);
    });
  },

  _onSelectAllIconsClicked: function (event) {
    event.preventDefault();
    this._iconCollection.each(function (icon) {
      if (icon.get('visible')) {
        icon.set('selected', true);
      }
    });
  },

  _onDeleteIconsClicked: function (event) {
    event.preventDefault();
    this._openDeleteIconsDialog();
  },

  _bindDeleteIconsDialog: function () {
    cdb.god.bind('closeDialogs:delete', this._destroyDeleteIconsDialog, this);
  },

  _unbindDeleteIconsDialog: function () {
    cdb.god.unbind('closeDialogs:delete', this._destroyDeleteIconsDialog, this);
  },

  _openDeleteIconsDialog: function (e) {
    this.killEvent(e);

    cdb.god.trigger('closeDialogs:delete');

    this.delete_icons_dialog = new DeleteIconsDialog({
      numOfIcons: this._getNumberOfSelectedIcons(),
      okCallback: this._deleteIcons.bind(this)
    });

    this.delete_icons_dialog.appendToBody();

    this._bindDeleteIconsDialog();
    this.addView(this.delete_icons_dialog);
  },

  _deleteIcons: function () {
    this._hideErrorMessage();
    var iconsToDelete = this._iconCollection.where({ selected: true });

    _.each(iconsToDelete, function (icon) {
      icon.destroy({
        beforeSend: this._beforeIconDelete.bind(this),
        success: this._onIconDeleted.bind(this),
        error: this._onIconDeleteError.bind(this),
        complete: this._onIconDeleteComplete.bind(this)
      });
    }, this);
  },

  _destroyDeleteIconsDialog: function () {
    if (this.delete_icons_dialog) {
      this._unbindDeleteIconsDialog();
      this.delete_icons_dialog.remove();
      this.removeView(this.delete_icons_dialog);
      this.delete_icons_dialog.hide();
      delete this.delete_icons_dialog;
    }
  },

  _getNumberOfSelectedIcons: function () {
    return this._iconCollection.where({ selected: true }).length;
  },

  _beforeIconDelete: function () {
    this._numOfDeletingProcesses++;
    if (this._numOfDeletingProcesses > 0) {
      this._runningMessage = 'Deleting icons...';
      this.model.set('isProcessRunning', true);
    }
  },

  _onIconDeleted: function (icon) {
    icon.set('deleted', true);
    this._addExtraIcon();
  },

  _onIconDeleteError: function (icon, response) {
    var errorText = this._parseResponseText(response);
    // Even if API throws error the icon has been already removed from the collection.
    // We must add it again
    this._iconCollection.add(icon);
    this._resetSelection();
    this._showErrorMessage(this._deleteErrorMessage(errorText));
  },

  _onIconDeleteComplete: function () {
    this._numOfDeletingProcesses--;
    if (this._numOfDeletingProcesses <= 0) {
      this.model.set('isProcessRunning', false);
    }
  },

  _showSpinner: function (enable) {
    if (enable) {
      this._hide('.js-plusSign');
      this._show('.js-spinner');
    } else {
      this._show('.js-plusSign');
      this._hide('.js-spinner');
    }
  },

  _showErrorMessage: function (message) {
    this.$('.js-errorMessage label').text(message);
    this._show('.js-errorMessage');
  },

  _hideErrorMessage: function () {
    this._hide('.js-errorMessage');
  },

  _resetSelection: function () {
    this._iconCollection.each(function (icon) {
      icon.set('selected', false);
    });
    this._refreshActions();
  },

  _addExtraIcon: function () {
    var iconAdded = false;
    this._iconCollection.each(function (icon) {
      if (!icon.get('visible') && !iconAdded) {
        this._addIconElement(icon);
        iconAdded = true;
      }
    }, this);
  },

  _onProcessRunningChanged: function () {
    var running = this.model.get('isProcessRunning');
    this._showSpinner(running);
    if (running) {
      this.$el.css('pointer-events', 'none');
      this._hideActions();
      this.$('.js-runningInfo').text(this._runningMessage);
      this._show('.js-runningInfo');
    } else {
      this.$('.js-runningInfo').text('');
      this._hide('.js-runningInfo');
      this._refreshActions();
      cdb.god.trigger('refreshCollection', { cid: this.cid });
      this.$el.css('pointer-events', 'auto');
    }
  },

  _uploadErrorMessage: function (errorText) {
    var message = 'Error uploading your image. ';
    if (errorText) {
      message += '[ ' + errorText + ' ]. ';
    }
    message += 'Please try again.';

    return message;
  },

  _deleteErrorMessage: function (errorText) {
    var message = 'Error deleting your image. ';
    if (errorText) {
      message += '[ ' + errorText + ' ]. ';
    }
    message += 'Please try again.';

    return message;
  }
});
