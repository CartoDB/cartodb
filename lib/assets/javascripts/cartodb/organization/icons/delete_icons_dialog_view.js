var BaseDialog = require('../../common/views/base_dialog/view');

module.exports = BaseDialog.extend({

  initialize: function () {
    if (!this.options.okCallback || typeof this.options.okCallback !== 'function') {
      throw new Error('Callback function for OK action is mandatory.');
    }
    this.elder('initialize');
    this._numOfIcons = this.options.numOfIcons || 0;
    this._okCallback = this.options.okCallback;
  },

  render_content: function () {
    return this.getTemplate('organization/icons/delete_icons_modal')({
      numOfIcons: this._numOfIcons
    });
  },

  ok: function () {
    this._okCallback();
    this.close();
  }
});
