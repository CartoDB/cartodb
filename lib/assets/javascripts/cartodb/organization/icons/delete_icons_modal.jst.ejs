<div class="CDB-Text Dialog-header u-inner">
  <div class="Dialog-headerIcon Dialog-headerIcon--negative">
    <i class="CDB-IconFont CDB-IconFont-trash"></i>
    <% if (numOfIcons > 0) { %>
    <span class="Badge Badge--negative Dialog-headerIconBadge CDB-Text CDB-Size-small"><%- numOfIcons %></span>
    <% } %>
  </div>
  <h4 class="CDB-Text CDB-Size-large u-mainTextColor u-secondaryTextColor u-bSpace--m u-tSpace-xl">
    You are about to delete
    <% if (numOfIcons <= 0) { %>
      icons 
    <% } else if (numOfIcons === 1) { %>
      1 icon 
    <% } else { %>
      <%- numOfIcons  %> icons
    <% } %>
    from your organization.
  </h4>
  <p class="CDB-Text CDB-Size-medium u-altTextColor">
    Maps using
    <% if (numOfIcons <= 1) { %>
      this icon
    <% } else { %>
      these icons
    <% } %>
    will be affected.
  </p>
  <p class="CDB-Text CDB-Size-medium u-altTextColor">
    Also, the deleted 
    <% if (numOfIcons === 1) { %>
      icon
    <% } else { %>
      icons 
    <% } %>
    cannot be recovered.
  </p>
  <p class="CDB-Text CDB-Size-medium u-altTextColor">
    Be sure before proceeding.
  </p>
</div>

<div class="Dialog-footer Dialog-footer--simple u-inner">
  <div class="Dialog-footerContent">
    <button class="CDB-Button CDB-Button--secondary cancel">
      <span class="CDB-Button-Text CDB-Text is-semibold CDB-Size-small u-upperCase">Cancel</span>
    </button>
    <button class="u-lSpace--xl CDB-Button CDB-Button--error ok">
      <span class="CDB-Button-Text CDB-Text is-semibold CDB-Size-small u-upperCase">Ok, delete</span>
    </button>
  </div>
</div>
