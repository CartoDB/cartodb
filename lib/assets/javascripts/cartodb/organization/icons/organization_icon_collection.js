var cdb = require('cartodb.js-v3');
var Backbone = require('backbone-cdb-v3');
var IconModel = require('./organization_icon_model');

module.exports = Backbone.Collection.extend({

  model: IconModel,

  url: function (method) {
    var version = cdb.config.urlVersion('organization-assets', method);
    return '/api/' + version + '/organization/' + this.orgId + '/assets';
  },

  initialize: function (attrs, opts) {
    if (!opts.orgId) {
      throw new Error('Organization ID is required');
    }
    this.orgId = opts.orgId;
  }
});
