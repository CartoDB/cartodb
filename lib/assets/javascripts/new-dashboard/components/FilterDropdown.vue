<template>
<section v-click-outside="closeDropdown">
  <button class="dropdown__toggle" @click="toggleDropdown" :class="{ 'dropdown__toggle--active': isDropdownOpen }">
    <slot />
  </button>

  <div class="dropdown" :class="{ 'is-open': isDropdownOpen }">
    <div class="section">
      <h6 class="text is-xsmall is-txtSoftGrey u-tupper letter-spacing">{{ $t('FilterDropdown.showMe') }}</h6>
      <ul class="list">
        <li class="type text is-caption is-txtGrey" :class="{ 'type--selected': isFilterApplied('mine') }">
          <a href="javascript:void(0)" class="element" :class="{ 'element--selected': isFilterApplied('mine') }" @click="setFilter('mine')">
            {{ $t(`FilterDropdown.types.${section}`, { count: metadata.total_user_entries }) }}
          </a>
        </li>
        <li class="type text is-caption is-txtGrey" :class="{ 'type--selected': isFilterApplied('shared') }" v-if="metadata.total_shared">
          <a href="javascript:void(0)" class="element" :class="{ 'element--selected': isFilterApplied('shared') }" @click="setFilter('shared')">
            {{ $t('FilterDropdown.types.shared', { count: metadata.total_shared }) }}
          </a>
        </li>
        <li class="type text is-caption is-txtGrey" :class="{ 'type--selected': isFilterApplied('favorited') }" v-if="metadata.total_likes">
          <a href="javascript:void(0)" class="element" :class="{ 'element--selected': isFilterApplied('favorited') }" @click="setFilter('favorited')">
            {{ $t('FilterDropdown.types.favorited', { count: metadata.total_likes }) }}
          </a>
        </li>
        <li class="type text is-caption is-txtGrey" :class="{ 'type--selected': isFilterApplied('locked') }">
          <a href="javascript:void(0)" class="element" :class="{ 'element--selected': isFilterApplied('locked') }" @click="setFilter('locked')">
            {{ $t('FilterDropdown.types.locked') }}
          </a>
        </li>
        <li class="type text is-caption is-txtGrey" :class="{ 'type--selected': isPrivacyFilterApplied }">
          <a href="javascript:void(0)" class="element" :class="{ 'element--selected': isFilterApplied('public') }" @click="setFilter('public')">
            {{ $t('FilterDropdown.types.publicPrivacy') }}
          </a> |
          <a href="javascript:void(0)" class="element" :class="{ 'element--selected': isFilterApplied('private') }" @click="setFilter('private')">
            {{ $t('FilterDropdown.types.privatePrivacy') }}
          </a> |
          <a href="javascript:void(0)" class="element" :class="{ 'element--selected': isFilterApplied('link') }" @click="setFilter('link')">
            {{ $t('FilterDropdown.types.linkPrivacy') }}
          </a> |
          <a href="javascript:void(0)" class="element" :class="{ 'element--selected': isFilterApplied('password') }" @click="setFilter('password')">
            {{ $t('FilterDropdown.types.passwordPrivacy') }}
          </a>
        </li>
      </ul>
    </div>
    <div class="section">
      <h6 class="text is-xsmall is-txtSoftGrey u-tupper letter-spacing">{{ $t('FilterDropdown.orderMaps') }}</h6>
      <ul class="list">
        <li class="type text is-caption is-txtGrey" :class="{ 'type--selected': isOrderApplied('favouritesFirst') }">
          <a href="javascript:void(0)" class="element" :class="{ 'element--selected': isOrderApplied('favouritesFirst') }" @click="setOrder('favouritesFirst')">
            {{ $t('FilterDropdown.order.favourites') }}
          </a>
        </li>
        <li class="type text is-caption is-txtGrey">
          {{ $t('FilterDropdown.order.alphabetical.title') }} (
            <a href="javascript:void(0)" class="element" @click="setOrder('alphabetically')">
              {{ $t('FilterDropdown.order.alphabetical.A-Z') }}
            </a> |
            <a href="javascript:void(0)" class="element" @click="setOrder('alphabeticallyReverse')">
              {{ $t('FilterDropdown.order.alphabetical.Z-A') }}
            </a>
          )
        </li>
        <li class="type text is-caption is-txtGrey" :class="{ 'type--selected': isOrderApplied('updated_at') }">
          Date Modified (
            <a href="javascript:void(0)" class="element" :class="{ 'element--selected': isOrderApplied('updated_at') }" @click="setOrder('updated_at')">
              {{ $t('FilterDropdown.order.date.last') }}
            </a> |
            <a href="javascript:void(0)" class="element">
              {{ $t('FilterDropdown.order.date.first') }}
            </a>
          )
        </li>
      </ul>
    </div>
  </div>
</section>
</template>

<script>
export default {
  name: 'MapCard',
  props: {
    section: String,
    filter: String,
    order: String,
    metadata: {
      type: Object,
      default () {
        return { total_shared: 0 };
      }
    }
  },
  data: function () {
    return {
      isVisible: false,
      isDropdownOpen: false
    };
  },
  computed: {
    isPrivacyFilterApplied () {
      return ['public', 'private', 'link', 'password'].indexOf(this.$props.filter) > -1;
    }
  },
  methods: {
    isOrderApplied (order) {
      return this.$props.order === order;
    },
    isFilterApplied (filter) {
      return this.$props.filter === filter;
    },
    toggleDropdown () {
      this.isDropdownOpen = !this.isDropdownOpen;
    },
    closeDropdown () {
      this.isDropdownOpen = false;
    },
    setFilter (filter) {
      this.closeDropdown();
      this.$emit('filterChanged', filter);
    },
    setOrder (order) {
      this.closeDropdown();
      this.$emit('orderChanged', order);
    }
  }
};
</script>

<style scoped lang="scss">
@import 'stylesheets/new-dashboard/variables';

.dropdown {
  visibility: hidden;
  position: absolute;
  z-index: 2;
  right: 0;
  width: 310px;
  margin-top: 8px;
  overflow: hidden;
  transition: all 0.25s linear;
  border: 1px solid $light-grey;
  border-radius: 2px;
  opacity: 0;
  pointer-events: none;

  &.is-open {
    visibility: visible;
    opacity: 1;
    pointer-events: initial;
  }
}

.section {
  padding: 24px 12px 24px 36px;
  border-bottom: 1px solid $light-grey;
  background-color: #FFF;

  &:last-of-type {
    border-bottom: none;
  }

  &:nth-child(2n) {
    background-color: $softblue;
  }
}

.list {
  margin-top: 8px;
}

.type {
  margin-bottom: 8px;

  &:last-of-type {
    margin-bottom: 0;
  }

  &.type--selected {
    position: relative;

    &::before {
      content: "";
      position: absolute;
      top: 50%;
      left: -22px;
      width: 14px;
      height: 14px;
      transform: translateY(-50%);
      background-image: url("../assets/icons/common/check.svg");
      background-repeat: no-repeat;
      background-position: center;
    }
  }
}

.element {
  text-decoration: none;

  &.element--selected {
    color: $text-color;
    pointer-events: none;
  }
}

.dropdown__toggle {
  display: flex;
  align-items: center;
  justify-content: center;
  height: 36px;
  padding: 0 9px;
  border-radius: 2px;

  &.dropdown__toggle--active {
    background-color: #F2F6F9;
  }
}
</style>
