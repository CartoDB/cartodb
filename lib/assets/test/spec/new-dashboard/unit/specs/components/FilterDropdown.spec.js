import { shallowMount } from '@vue/test-utils';
import FilterDropdown from 'new-dashboard/components/FilterDropdown';
import { filterIcon } from 'icons';

const $t = key => key;

describe('FilterDropdown.vue', () => {
  it('should render correct contents based on filter and order', () => {
    const filterDropdown = createFilterDropdown({
      filter: 'link',
      order: 'updated_at',
      section: 'maps',
      metadata: {
        total_likes: 10,
        total_shared: 10
      }
    });
    expect(filterDropdown).toMatchSnapshot();
  });

  it('should open dropdown when clicking toggle', () => {
    const filterDropdown = createFilterDropdown();

    filterDropdown.find('.dropdown__toggle').trigger('click');

    expect(filterDropdown).toMatchSnapshot();
  });

  describe('Methods', () => {
    describe('setFilter', () => {
      const filterDropdown = createFilterDropdown();
      const filterElement = filterDropdown.findAll('.element').at(1);

      filterElement.trigger('click');

      expect(filterDropdown.emitted('filterChanged')).toBeTruthy();
      expect(filterDropdown.emitted().filterChanged[0]).toEqual(['shared']);
    });

    describe('setOrder', () => {
      const filterDropdown = createFilterDropdown();
      const filterElement = filterDropdown.find('.section:nth-child(2) .element');

      filterElement.trigger('click');

      expect(filterDropdown.emitted('orderChanged')).toBeTruthy();
      expect(filterDropdown.emitted().orderChanged[0]).toEqual(['favouritesFirst']);
    });
  });
});

function createFilterDropdown (
  propsData = {
    section: 'maps',
    filter: 'mine',
    order: 'updated_at',
    metadata: { total_likes: 10, total_shared: 10 }
  }) {
  return shallowMount(FilterDropdown, {
    mocks: { $t },
    propsData,
    slots: {
      default: filterIcon
    }
  });
}
