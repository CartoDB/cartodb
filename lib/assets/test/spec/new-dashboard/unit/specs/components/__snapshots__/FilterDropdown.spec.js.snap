// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterDropdown.vue should open dropdown when clicking toggle 1`] = `
<section><button class="dropdown__toggle dropdown__toggle--active"><svg width="18" height="20" viewBox="0 0 18 20" xmlns="http://www.w3.org/2000/svg">
      <g fill="#036FE2" fill-rule="evenodd">
        <path d="M8.3 6.7l1.4-1.4L5 .58.3 5.29l1.4 1.42L4 4.4v11.6h2V4.4zM16.3 13.3L14 15.58V4h-2V15.6l-2.3-2.3-1.4 1.42 4.7 4.7 4.7-4.7z"></path>
      </g>
    </svg></button>
  <div class="dropdown is-open">
    <div class="section">
      <h6 class="text is-xsmall is-txtSoftGrey u-tupper letter-spacing">FilterDropdown.showMe</h6>
      <ul class="list">
        <li class="type text is-caption is-txtGrey type--selected"><a href="javascript:void(0)" class="element element--selected">
            FilterDropdown.types.maps
          </a></li>
        <li class="type text is-caption is-txtGrey"><a href="javascript:void(0)" class="element">
            FilterDropdown.types.shared
          </a></li>
        <li class="type text is-caption is-txtGrey"><a href="javascript:void(0)" class="element">
            FilterDropdown.types.favorited
          </a></li>
        <li class="type text is-caption is-txtGrey"><a href="javascript:void(0)" class="element">
            FilterDropdown.types.locked
          </a></li>
        <li class="type text is-caption is-txtGrey"><a href="javascript:void(0)" class="element">
            FilterDropdown.types.publicPrivacy
          </a> |
          <a href="javascript:void(0)" class="element">
            FilterDropdown.types.privatePrivacy
          </a> |
          <a href="javascript:void(0)" class="element">
            FilterDropdown.types.linkPrivacy
          </a> |
          <a href="javascript:void(0)" class="element">
            FilterDropdown.types.passwordPrivacy
          </a></li>
      </ul>
    </div>
    <div class="section">
      <h6 class="text is-xsmall is-txtSoftGrey u-tupper letter-spacing">FilterDropdown.orderMaps</h6>
      <ul class="list">
        <li class="type text is-caption is-txtGrey"><a href="javascript:void(0)" class="element">
            FilterDropdown.order.favourites
          </a></li>
        <li class="type text is-caption is-txtGrey">
          FilterDropdown.order.alphabetical.title (
          <a href="javascript:void(0)" class="element">
            FilterDropdown.order.alphabetical.A-Z
          </a> |
          <a href="javascript:void(0)" class="element">
            FilterDropdown.order.alphabetical.Z-A
          </a>
          )
        </li>
        <li class="type text is-caption is-txtGrey type--selected">
          Date Modified (
          <a href="javascript:void(0)" class="element element--selected">
            FilterDropdown.order.date.last
          </a> |
          <a href="javascript:void(0)" class="element">
            FilterDropdown.order.date.first
          </a>
          )
        </li>
      </ul>
    </div>
  </div>
</section>
`;

exports[`FilterDropdown.vue should render correct contents based on filter and order 1`] = `
<section><button class="dropdown__toggle"><svg width="18" height="20" viewBox="0 0 18 20" xmlns="http://www.w3.org/2000/svg">
      <g fill="#036FE2" fill-rule="evenodd">
        <path d="M8.3 6.7l1.4-1.4L5 .58.3 5.29l1.4 1.42L4 4.4v11.6h2V4.4zM16.3 13.3L14 15.58V4h-2V15.6l-2.3-2.3-1.4 1.42 4.7 4.7 4.7-4.7z"></path>
      </g>
    </svg></button>
  <div class="dropdown">
    <div class="section">
      <h6 class="text is-xsmall is-txtSoftGrey u-tupper letter-spacing">FilterDropdown.showMe</h6>
      <ul class="list">
        <li class="type text is-caption is-txtGrey"><a href="javascript:void(0)" class="element">
            FilterDropdown.types.maps
          </a></li>
        <li class="type text is-caption is-txtGrey"><a href="javascript:void(0)" class="element">
            FilterDropdown.types.shared
          </a></li>
        <li class="type text is-caption is-txtGrey"><a href="javascript:void(0)" class="element">
            FilterDropdown.types.favorited
          </a></li>
        <li class="type text is-caption is-txtGrey"><a href="javascript:void(0)" class="element">
            FilterDropdown.types.locked
          </a></li>
        <li class="type text is-caption is-txtGrey type--selected"><a href="javascript:void(0)" class="element">
            FilterDropdown.types.publicPrivacy
          </a> |
          <a href="javascript:void(0)" class="element">
            FilterDropdown.types.privatePrivacy
          </a> |
          <a href="javascript:void(0)" class="element element--selected">
            FilterDropdown.types.linkPrivacy
          </a> |
          <a href="javascript:void(0)" class="element">
            FilterDropdown.types.passwordPrivacy
          </a></li>
      </ul>
    </div>
    <div class="section">
      <h6 class="text is-xsmall is-txtSoftGrey u-tupper letter-spacing">FilterDropdown.orderMaps</h6>
      <ul class="list">
        <li class="type text is-caption is-txtGrey"><a href="javascript:void(0)" class="element">
            FilterDropdown.order.favourites
          </a></li>
        <li class="type text is-caption is-txtGrey">
          FilterDropdown.order.alphabetical.title (
          <a href="javascript:void(0)" class="element">
            FilterDropdown.order.alphabetical.A-Z
          </a> |
          <a href="javascript:void(0)" class="element">
            FilterDropdown.order.alphabetical.Z-A
          </a>
          )
        </li>
        <li class="type text is-caption is-txtGrey type--selected">
          Date Modified (
          <a href="javascript:void(0)" class="element element--selected">
            FilterDropdown.order.date.last
          </a> |
          <a href="javascript:void(0)" class="element">
            FilterDropdown.order.date.first
          </a>
          )
        </li>
      </ul>
    </div>
  </div>
</section>
`;
